README.md
[![Build Status](https://api.travis-ci.org/gpdviz/gpdviz.svg?branch=master)](https://travis-ci.org/gpdviz/gpdviz)

# Gpdviz

Gpdviz is a generic, lightweight tool for web-based
visualization of geo-located point data streams in real-time.

**Status**: Functional but still considered experimental.

Data providers use the Gpdviz REST API to register sensor systems,
data streams, and observations.
This API is specified using [OpenAPI 2.0](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md).
The visualization is updated in real-time using WebSockets.

Assuming Gpdviz is deployed at `http://example.net/gpdviz`:

| url | purpose |
| --- | --- |
| `http://example.net/gpdviz/mysysid/` | Visualization of sensor system with ID `mysysid`
| `http://example.net/gpdviz/api`      | The REST API endpoint
| `http://example.net/gpdviz/api-docs` | API documentation


**Data model**

The main entities in the data model are:
_sensor systems_, _data streams_, _variable definitions_, and _observations_.

- A _sensor system_ has associated metadata (name, description, etc.)
  and consists of a set of _data streams_.

- A _data stream_ has associated metadata (name, description, map/chart styles, etc.)
  and consists of a set of _variable definitions_ and associated _observations_.

- A _variable definition_ consists of a name, units, and chart style.

- An _observation_ is timestamped and can capture scalar data (e.g., temperature values),
  as well as features and geometries (in GeoJson format).
 
**Operations for data providers**

- Register/unregister sensor systems
- Add/remove data streams
- Add/remove variable definitions
- Register observations

The Gpdviz server maintains a registry with the provided information for each sensor system
including a window of most recent reported observations for each stream in the system.
Users will be able to see the latest reported state whenever they open the particular
URL corresponding to the desired sensor system, and in real time see any subsequent
updates reported by the provider.


## Running

### Direct execution

To execute Gpdviz you will need a [Java runtime environment](https://www.java.com/),
access to a PostgreSQL server, and ability to create a database and a user on that
server for Gpdviz purposes.

Download the latest executable JAR `gpdviz-x.y.z.jar` from https://github.com/gpdviz/gpdviz/releases/.

- Gpdviz expects a number of parameters for its regular execution. These parameters are to be
  indicated in a local `conf/gpdviz.conf` file. A template of such file, with
  a description of the various parameters, can be generated as follows:

        $ java -jar gpdviz-x.y.z.jar generate-conf

- Edit `conf/gpdviz.conf` as needed.

- Create the database and user indicated in `gpdviz.conf`, with the user being granted
  all privileges on the database.

- Execute the Gpdviz server:

        $ java -jar gpdviz-x.y.z.jar run-server


### Docker

With a Docker engine on your target system, you will only need to define some
parameters (via environment variables) and launch the services defined in
[docker-compose.yml](docker/docker-compose.yml).
This set-up includes the PostgreSQL server and database.
Please see https://hub.docker.com/r/gpdviz/gpdviz/.

### Demo

This [httpie](https://httpie.io/)-based
[bash script](data/ss1.demo.sh)
can be used as a demo of a data provider. Define the `GPDVIZ_SERVER` environment
variable prior to running this script.

For a similar demo in Python, see https://github.com/gpdviz/gpdviz_python_client_example,
which uses the client module https://github.com/gpdviz/gpdviz_python_client
automatically generated from the OpenAPI specification.

## Development

Gpdviz implementation is based on:

- [Akka-HTTP](https://doc.akka.io/docs/akka-http/current/)
- [Slick](https://github.com/slick/slick)
- [PostgreSQL](https://github.com/postgres/postgres)
- [ScalaJS](https://www.scala-js.org/)
- [Leaflet](https://leafletjs.com/)
- [Highcharts](https://www.highcharts.com/)


### Build and run

    $ npm install jsdom source-map-support
	$ sbt
	> package   # to copy js resources to jvm's classpath needed for the webapp
	> gpdvizJVM/runMain gpdviz.Gpdviz generate-conf
	> gpdvizJVM/runMain gpdviz.Gpdviz create-tables
	> gpdvizJVM/runMain gpdviz.Gpdviz add-some-data
	> gpdvizJVM/runMain gpdviz.Gpdviz run-server

For the demo with sensor system ID "ss1":

- Open http://localhost:5050/ss1/ in your browser.
- On another terminal, either run the bash script:

	    $ data/ss1.demo.sh
	
    or the similar client in python as explained at
    https://github.com/gpdviz/gpdviz_python_client_example


![](static/gpdviz2.gif)


### Dist

To generate the executable JAR `jvm/target/scala-2.12/gpdviz-x.y.z.jar`:

	$ sbt gpdvizJVM/assembly

For the dockerized version, see [docker/readme.md](docker/readme.md).

### Model

Data model details are still wip.

### API

OpenAPI spec generated by [swagger-akka-http](https://github.com/swagger-akka-http/swagger-akka-http).

**Swagger UI**

A Swagger UI instance is embedded as follows:

- copy [Swagger UI dist/*](https://github.com/swagger-api/swagger-ui/tree/master/dist)
  contents to `jvm/src/main/resources/swaggerui/`;
- adjust the `url` entry in `swaggerui/index.html` to refer to the generated spec:
  `url: window.location + "/swagger.json"`
- rebuild

Under local development:
- generated spec at http://localhost:5050/api-docs/swagger.json
- Swagger UI at http://localhost:5050/api-docs



## history

[Original idea](https://github.com/carueda/gpdviz0).


ChangeLog.md
# Gpdviz

2022-01

- TODO: focus this project on the backend, not frontend.
  Use something like [scala-tsi](https://github.com/scala-tsi/scala-tsi)
  to help capture the data models for the UI.
- TODO: overall cleanup and upgrade to scala3

- fix data from `add-some-data`

2021-10

- minor adjustments while revisiting this project.
  Adding some old, previously untracked files (under misc/)/

2018-11-13

- minor docker adjustments to run this again on my local machine:
    - move docker-compose.yml to `docker/`
    - edit `docker/setenv.sh` to adjust couple paths
- minor: `restart: always` in docker-compose

2018-07-10 0.4.5

- capture and handle zoom in POST and PUT sensor system
- publish docker image gpdviz/gpdviz:0.4.5

2018-06-27 0.4.4

- some minor docker notes
- fix in PositionsByTime.get to actually get position of very latest
  element (when that's the case of course) instead of the previous one
- for now, include circle for selected point (useful for front tracking)
  TODO make this optional via some map attribute
- remember last given point for a more efficient addSelectionPoint

2018-06-04 0.4.3

- publish new docker image:

        sbt gpdvizJVM/assembly
        docker/build.sh gpdviz
        docker push gpdviz/gpdviz:0.4.3

- quick "fix" for upcoming exercise with 'salinity':
  along with 'temperature' also check for 'salinity' for
  `seriesIndexForMouseHover` in charter.js

- really quick add "leaflet-measure"
  TODO cleanup as way too much not needed in version control

2017-12-19 0.4.2

- hide position marker on map when mouse goes outside of chart
  and when chart is closed

2017-11-10 0.4.1

- encodeResponse in some GET's
- include option for google map if new config property googleMapApiKey is given
- don't include pusher.js sensor system page if not configured to use it
- use Double for timeMs in PositionsByTime
- update akka-http to 10.0.10
- update highstock to 6.0.2
- update scala to 2.12.4
- update scalajs to 0.6.21
- update sbt to v1 (1.0.2)

2017-11-07 0.4.0

- docker images under gpdviz organization (https://hub.docker.com/u/gpdviz/)
- version source for all build stuff is now jvm/src/main/resources/reference.conf.
  At run time, when using dockerized image, the version is only to be indicated
  explicitly in docker-compose.yml.
- complete initial working dockerization
- create tables in run-server if nonexistent

2017-11-02

- pass ExecutionContext in some places.
  ("The recommended approach is to add `(implicit ec: ExecutionContext)` to methods,
   or class constructor parameters, which need an `ExecutionContext`.)
- re-enable event notifications depending on sensor system's pushEvents flag.
- Added MockPublisher to test the notifications

2017-10-31

- observation timestamp:
  - internally captured as timestamptz in database
  - externally required and exposed as ISO-8601 formatted string

2017-10-30
- use OffsetDateTime for time in observation table.
- use PgSprayJsonSupport
- adjust swagger spec
- fix update sensor system
- remove PostgresDbDoobie and PostgresDbQuill
- route to add variable definitions
- adjust response for successful creation of entities
- overall register/unregister-to-add/delete renaming
- add drop-tables, especially to avoid error noise during local testing

2017-10-25

- complete retrieval of full sensor system
- add option to import existing models in json files

2017-10-23

- first slick-based version

2017-10-20

- due to doobie issues, trying slick
- due to quill issues, trying again with doobie
- some db-related refact
- nested embedded with intermediate Option still not handled by quill

2017-10-19

- do not forget to also update the swagger response types!
- adjustments to API mainly in terms of responses (so they are all objects)
- tests re-enabled (except for one involving quill's update operation)

2017-10-14

- focus on quill/postgres as the database system.
  Some tests ignored for now.
  ss1.demo.sh working.
  Define config property `gpdviz.postgres` to exercise it.
  DB setup:

        psql -c 'create database gpdviz;' -U postgres
        psql -U postgres -d gpdviz < create_tables.sql

2017-09-14 0.3.2

- fix api-docs link
- update some doc

2017-09-13 0.3.1

- simplify execution via explicit `gpdviz.conf` file

2017-09-12 0.3.1

- add demo data entry generated from [gpdviz_python_client_example](https://github.com/gpdviz/gpdviz_python_client_example)
- repo moved to gpdviz organization
- swagger adjustments, in particular, use of `@(ApiModelProperty @field)(dataType = "object")`
  to annotate fields with type involving `JsObject`, `Feature`, or `Geometry`, which are not
  properly exposed; moreover, the `object` dataType gives good flexibility (for now).
- dispatch root request with list of registered systems

2017-09-01 0.3.0

- include link to api-docs
- capture externalUrl in configuration
- use reference.conf to expose version to app
- use swagger to generate openApi spec and api-docs ui
  - generated spec at `.../api-docs/swagger.json`, eg., http://localhost:5050/api-docs/swagger.json
  - Swagger UI at `.../api-docs/`, eg., http://localhost:5050/api-docs

2017-08-31

- some refact - separation of rest services
- expose tool version in web page
- preliminaries to enable support for mongo as db

2017-08-30

- make data interfaces non-blocking

2017-08-29

- apache config to support websocket per https://stackoverflow.com/a/29823699/830737
- various adjustments while testing on external servers and back in local mode.

2017-08-25

- version 0.2.0
- repo renamed to https://github.com/carueda/gpdviz.
  (old one to https://github.com/carueda/gpdviz0)
- publish websocket messages for the specific sysid
- simplify details panel
- restore window resize handling
- some investigation re "Websocket handler failed with Processor actor" when exiting the server
  while some client currently connected. TODO check https://stackoverflow.com/a/41359874/830737
- handle chart height and minWidth
- handle map center and zoom

data/ss1.demo.sh
#!/usr/bin/env bash

GPDVIZ="${GPDVIZ_SERVER:-http://localhost:5050}"

echo "Using gpdviz server = ${GPDVIZ}"

SS=ss1

baseTimeSec=1503090553

RANDOM=1341

function timeSec2iso() {
    secs=$1
    gdate --date="@${secs}" --iso-8601=seconds
}

function run() {
	unregister
	register

    generate_str1
    generate_str2
    generate_str3
    generate_str4
}

function unregister() {
    echo "unregister ${SS}"
	http delete ${GPDVIZ}/api/ss/${SS} > /dev/null
}

function register() {
    echo "register ${SS}"
	http post ${GPDVIZ}/api/ss sysid=${SS} \
	     name='Test sensor system' \
	     description='Description of Test sensor system' \
	     center:='{"lat":36.82, "lon":-122}' \
	     pushEvents:=true > /dev/null
}

function generate_str1() {
	add_str1
	now=${baseTimeSec}
	secs=60
	timeSec=$(( now - secs ))
	add_str1_polygon ${timeSec}
	add_scalars str1 $(( timeSec + 1 )) $(( secs - 1 ))
}

function add_str1() {
    echo "add stream str1 (with baz variable)"
	http post ${GPDVIZ}/api/ss/${SS} strid=str1 \
	     name="Stream one" \
	     description="Description of Stream one" \
	     chartStyle:='{
	       "useChartPopup": true,
	       "height": 500,
            "yAxis": [{
	             "height": "50%",
	             "title": {"text": "baz (m)"},
	             "opposite": false,
	             "offset": -10
	           }, {
	             "top": "55%",
	             "height": "45%",
	             "title": {"text": "temperature (°)"},
	             "opposite": false,
	             "offset": -10
            }]
	     }' \
	     variables:='[
	     {
	       "name": "baz",
	       "units": "m",
	       "chartStyle": {
	         "yAxis": 0,
	         "type": "column"
	       }
	     }]' \
	     mapStyle:='{"color":"green", "dashArray": "5,5"}' > /dev/null

    echo "add temperature variable to stream str1"
	http post ${GPDVIZ}/api/ss/${SS}/str1/vd \
	       name="temperature" \
	       units="°" \
	       chartStyle:='{
	         "yAxis": 1
	       }'
}

function add_str1_polygon() {
	timeSec=$1
    read -r -d '' geometry <<-EOF
      "`timeSec2iso ${timeSec}`": [{
        "geometry": {
          "type": "Polygon",
          "coordinates": [
            [[-121.8564,36.9], [-122.2217,36.9574], [-122.0945,36.6486], [-121.8674,36.6858]]
          ]
        }
      }]
EOF
    observations="{${geometry}}"
    add_observations str1 "${observations}"
}

function generate_str2() {
	add_str2
	
	secs=30
	timeSec=$(( timeSec - secs ))
	
	# geometries:
	read -r -d '' geometry <<-EOF
	  "`timeSec2iso ${timeSec}`": [
	    {
		  "geometry": {
			"type": "Point",
			"coordinates": [-121.906,36.882]
		  }
		}, {
		  "feature": {
			"properties": {
			  "style": {"color":"cyan", "radius": 20, "dashArray": "5,5"}
			},
			"geometry": {
			  "type": "Point",
			  "coordinates": [-121.965,36.81]
			}
		  }
		}]
EOF
    observations="{${geometry}}"
    add_observations str2 "${observations}"
	
	# data:
	echo "chart data: str2"
	observations='{'
	comma=""
	for i in `seq ${secs}`; do
		timeSec=$(( timeSec + 1 ))
		val0=$(( (RANDOM % 100) + 1 ))
		val1=$(( (RANDOM % 100) + 1 ))
		
        read -r -d '' element <<-EOF
          "`timeSec2iso ${timeSec}`": [{
            "scalarData": {
              "vars": ["foo", "bar"],
              "vals": [${val0}, ${val1}]
            }}]
EOF
		observations="${observations}${comma}${element}"
		comma=","
	done
	observations="${observations}}"
	add_observations str2 "${observations}"
}

function add_str2() {
    echo "add stream str2"
	http post ${GPDVIZ}/api/ss/${SS} strid=str2 \
	     variables:='[ { "name": "foo"}, {"name": "bar" } ]' \
	     mapStyle:='{"color":"red", "radius": 14}' zOrder:=10 > /dev/null
}

function generate_str3() {
	add_str3

	timeSec=${baseTimeSec}
    read -r -d '' geometry <<-EOF
      "`timeSec2iso ${timeSec}`": [{
        "geometry": {
			"type": "LineString",
			"coordinates": [
			    [-122.123,36.92], [-122.186,36.774], [-121.9,36.7]
			]
        }
      }]
EOF
    observations="{${geometry}}"
    add_observations str3 "${observations}"
}

function add_str3() {
    echo "add stream str3"
	http post ${GPDVIZ}/api/ss/${SS} strid=str3 \
	     mapStyle:='{"color":"blue"}' > /dev/null
}

function generate_str4() {
    strid=str4
	http post ${GPDVIZ}/api/ss/${SS} strid=${strid} \
	    name="${strid}_name" \
	    variables:='[ { "name": "temperature", "units": "°C" } ]' \
	    mapStyle:='{"color":"yellow", "radius": 10}' zOrder:=10 > /dev/null

	timeSec=${baseTimeSec}
    read -r -d '' geometry <<-EOF
      "`timeSec2iso ${timeSec}`": [{
          "geometry": {
            "type": "Point",
            "coordinates": [-122.09,36.865]
          }
      }]
EOF
    observations="{${geometry}}"
    add_observations "${strid}" "${observations}"
    echo "added ${strid} point: timeSec=${timeSec}"
	
	add_delayed_data ${strid} temperature 10
}

function add_delayed_data() {
    strid=$1
    varName=$2
	secs=$3
    timeSec=${baseTimeSec}
	for i in `seq ${secs}`; do
        sleep 1
	    timeSec=$(( timeSec + 1 ))
		val=$(( (RANDOM % 100) + 1 ))
        read -r -d '' element <<-EOF
          "`timeSec2iso ${timeSec}`": [{
            "scalarData": {
              "vars": ["${varName}"],
              "vals": [${val}]
            }}]
EOF
        observations="{${element}}"
        add_observations "${strid}" "${observations}"
	    echo "added observation to ${strid}: timeSec=${timeSec}"  # ${observations}"
	done
}

function add_observations() {
	strid=$1
	observations=$2
	http post ${GPDVIZ}/api/ss/${SS}/${strid}/obs \
	     observations:="${observations}" > /dev/null
}

function add_scalars() {
	strid=$1
	timeSec=$2
	secs=$3
	echo "scalarData: ${strid} timeSec=${timeSec}  secs=${secs}"
	observations='{'
	comma=""
	for i in `seq ${secs}`; do
		val0=$(( RANDOM % 1000 - 500 ))
		val1=$(( RANDOM % 100 ))
		lat="36.8$((   RANDOM % 1000 ))"
		lon="-122.1$(( RANDOM % 1000 ))"
    	read -r -d '' element <<-EOF
		  "`timeSec2iso ${timeSec}`": [{
            "scalarData": {
              "vars": ["baz","temperature"],
              "vals": [${val0}, ${val1}],
              "position": {"lat": ${lat}, "lon": ${lon}}
            }}]
EOF
		observations="${observations}${comma}${element}"
		comma=", "
		timeSec=$(( timeSec + 1 ))
	done
	observations="${observations}}"
	add_observations "${strid}" "${observations}"
}

if [ "$1" != "" ]; then
    $1 $2 $3 $4 $5
else
    run
fi
